src/arch/io/network.cc

```diff
@@ -621,6 +621,430 @@ void linux_tcp_conn_t::on_event(int /* events */) {
     event_watcher->stop_watching_for_errors();
 }
 
+/* tls_conn_wrapper_t wraps a TLS connection and safely releases it in
+exceptional cases if it is not reclaimed before execution leaves its scope. */
+class tls_conn_wrapper_t {
+public:
+    tls_conn_wrapper_t(SSL *_conn) {
```

Comment:
Also, since the connection is constructed inside the `tls_conn_wrapper_t` then, you'll need a `TLS *get()` method to get access to the pointer in the `linux_secure_tcp_conn_t`. On the plus side, it makes the code there simpler because you only need to keep the `tls_conn_wrapper_t` around and not both the wrapper and a pointer to the `conn` itself.